recommended architecture and deploy docs (#4795)
digital-asset/daml: docs/source/app-dev/app-arch.rst

.. Copyright (c) 2020 The DAML Authors. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Application architecture
########################

This section describes our recommended design of a full-stack DAML application.

> .. image:: ./recommended_architecture.svg

Reviewer: In the architecture image "External Service Provider" is crossed with an arrow. Would break the text so that it's not crossed

Author: fixed by moving the arrow origin one layer outside.

The presented architecture is by no means your only option, the DAML SDK provides support libraries
if you choose to develop your application in :ref:`Java <java-bindings>` and :ref:`Scala
<scala-bindings>`, or if you wish to develop directly against a :ref:`gRPC <grpc>` endpoint instead

> of the JSON REST API.

Reviewer: ```suggestion
The above image shows the recommended architecture. If you are working with a different tech stack DAML SDK also supports :ref:`Java <java-bindings>` and :ref:`Scala
<scala-bindings>` bindings as well as a direct :ref:`gRPC <grpc>` endpoint (instead of JSON API)
```

> To get you started quickly with the recommended application architecture, clone the

Reviewer: ```suggestion
To get started quickly with the recommended application architecture clone the```

``create-daml-app`` application template:

.. code-block:: bash

  git clone https://github.com/digital-asset/create-daml-app

``create-daml-app`` is a small, but fully functional demo application implementing the recommended
architecture, including

- DAML React libraries
- quick iteration against a sandbox backend
- deployment in the cloud or on `DABL <https://projectdabl.com>`_

> - authentication

Reviewer: ```suggestion
- DAML React libraries
- quick iteration against against the :ref:`DAML Ledger Sandbox<sandbox-manual>`
- authentication
- deployment in the cloud or on `DABL <https://projectdabl.com>`_
```
How is "deployment in the cloud" covered by the template? Do we provide scripts for automatically deploying to DABL and another cloud provider?

Author: it contains a docker file. we want to make the creation of the docker image a bit easier as a next step. also the part about deployment on DABL has yet to come.

Reviewer: Maybe then we should write something like this:

"DAR file that can be easily deployed to DABL as well as a Docker image that can be deployed to a cloud environment of your choice"

Author: changed it to `deploy your application in the cloud as a Docker container`. removed the part about dabl, let's wait until create-react-app actually contains instructions on how to deploy to dabl.

Author: also added the reference to the sandbox.

+==============================================================+==========================================================================+
| `@daml/react <https://www.npmjs.com/package/@daml/react>`_   | React hooks to query/create/exercise DAML contracts                      |
+--------------------------------------------------------------+--------------------------------------------------------------------------+
| `@daml/ledger <https://www.npmjs.com/package/@daml/ledger>`_ | DAML ledger object to connect and directly submit commands to the ledger |
+--------------------------------------------------------------+--------------------------------------------------------------------------+
| `@daml/types  <https://www.npmjs.com/package/@daml/types>`_  | DAML data types                                                          |
+--------------------------------------------------------------+--------------------------------------------------------------------------+

You can install any of these libraries by running ``yarn add @daml/react`` in the ``ui`` directory
of your project. Please explore the ``create-daml-app`` example project to see the usage of these

> libraries.

Reviewer: Can we please add examples how to use the library to the NPM for all mentioned packages? It would really make our documentation and developer experience much nicer than telling the user "go ahead and figure it out on your own". Our testers and internal members really complained when they had to "chace" it elsewhere

Author: do you want to have the doc on the npmjs.com packe or here? I thought I don't want to overload this section with details about those libraries here.

Reviewer: Not here 😄npmjs.com would be perfect

Author: there is actually already a bit of docs there, at least for @daml/react which you should mainly use.

Author: https://www.npmjs.com/package/@daml/react

Author: @daml/ledger is definitely missing some docs there.

Author: I will generate docs for the typescript libraries within the packages and add a section to the building `applications chapter` as a next step, but outside this PR.

The interface between the frontend and the backend of your application is defined by your DAML code.
To make your life easy when interacting with the ledger, the DAML assistant can generate
corresponding typescript data definitions for the data types declared in DAML.

.. code-block:: bash

>   daml codegen ts .daml/dist/<your-project-name.dar> -o daml2ts

Reviewer: We need to mention what's a dar file and reference the appropriate section in the docs, e.g., /concepts/glossary.html#dar-file-dalf-file

Author: I think that's more of a detail that you don't necessarily need to know at this point. I think it's more confusing than enlightening if you just want to get started.

This will create a typescript project under ``daml2ts`` that you can install by changing directory
to ``daml2ts`` and running ``yarn install``. We recommend using the provided ``daml-start.sh``
script at the root of the project directory of ``create-daml-app``. This will take care of creating

> and updating the ``daml2ts`` library whenever you change your DAML model automatically.

Reviewer: Are the two (yarn install and daml-start.sh) options or I need to do both, i.e., do I need to do yarn install and daml-start.sh? If I can use only create-daml-app/daml-start.sh I would only describe that case. Ideally this would be all done in the background (typescript codegen) when the user uses daml-start.sh.

Author: you need to do both, the process is not that smooth at this point.

If you choose a different frontend framework, the DAML assistant will also help you generate code

> for a variety of other programming languages. Run ``daml codegen --help`` to find out more.

Reviewer: If I run "daml codegen --help" it tells me that I can run it for scala, java, and ts. So I can't use it to generate	vue.js code or express.js code (frontend frameworks). Would remove this part and would describe that daml2ts provides data types for other frontend frameworks/libraries that could be used instead of React, e.g., Angular.

Author: changed it, not mentioning the other codegen's anymore.

Backend
~~~~~~~

The backend for your application can be any DAML ledger implementation. For applications where
performance is not of outmost importance, we recommend using the :ref:`DAML JSON API <json-api>` as
an interface. It is served by the HTTP JSON API server connected to the ledger API server. It
provides simple HTTP endpoints to interact with the ledger via GET/POST requests. If performance
does become a concern for you application, you can drop the HTTP JSON API server and use the
:ref:`gRPC API <grpc>` directly.

When you use the ``create-daml-app`` template application you can start a local sandbox together
with a JSON API server by running

.. code-block:: bash

  ./daml-start.sh

in the root of the project. This is the most simple DAML ledger implementation. Once your
application matures and becomes ready for production, the ``daml deploy`` command helps you deploy
your frontend and DAML artefacts of your project to a production ledger. See :ref:`Deploying to DAML

> Ledgers <deploy-ref_index>` for an in depth manual for specific ledgers.

Reviewer: deploy-ref_overview might work instead of deploy-ref_index (

Author: changed.

Authentication
~~~~~~~~~~~~~~

> DAML ledgers support a unified interface for authentication of commands, a detailed description can

Reviewer: Add a sentence why authentication is needed at the beginning of the section, e.g.,

To run your application against a :ref: `deployed ledger <deploy-ref_overview>`, you will need to add authentication.

Author: done.

offer an integrated authentication service, but you can also use an external service provider for
authentication like https://auth0.com. The DAML react libraries support interfacing with an
authenticated DAML ledger. Simply initialize your ``DamlLedger`` object with the token obtained by
an authentication service. How authentication works and the form of the required tokens is described
in the :ref:`Authentication <authentication>` section.

Developer workflow
~~~~~~~~~~~~~~~~~~

The DAML SDK enables a local development environment with fast iteration cycles. If you run
``daml-start.sh`` of the ``create-daml-app``, a local DAML sandbox ledger is started that is updated
with your most recent DAML code on any change. Next, you can start your frontend in development mode
by changing to your ``ui`` directory and run ``yarn start``. This will reload your frontend whenever
you make changes to it. You can add unit tests for your DAML models by writing :ref:`DAML scenarios
<testing-using-scenarios>`. These will also be reevaluated on change.
A typical DAML developer workflow is to

  #. Make a small change to your DAML data model

>   #. Optionally add a scenario unit test to assert correctness

Reviewer: ```suggestion
  #. Optionally test your DAML code and with scenarios (add reference to scenario testing in the docs)
```

>   #. Adapt the UI such that the code compiles again against the new DAML code

Reviewer: ```suggestion
  #. Edit your React components to be aligned with changes made in DAML code
```

Author: if you change your DAML code, your generated typescript datatypes will change and most likely your React code won't compile anymore.

  #. Extend the UI to make use of the newly introduced feature

>   #. Restart with step 1

Reviewer: ```suggestion
  #. Make further changes either to your DAML and/or React code until you're happy with what you've developed
```
Just wanna write something so that it doesn't sound like users need to do something, this step is optional 😄

.. Copyright (c) 2020 The DAML Authors. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

Application architecture
########################

This section describes our recommended design of a full-stack DAML application.

.. image:: ./recommended_architecture.svg

> The above image shows the recommended architecture. If you are working with a different tech stack DAML SDK also supports :ref:`Java <java-bindings>` and :ref:`Scala

Reviewer 2: I think it’s worth mentioning here that the JSON API can be used from any language and given that basically every dev is familiar with making HTTP requests this is a much smaller step than trying to use the gRPC API.

Author: agreed! I rearranged and added a sentence.

<scala-bindings>` bindings as well as a direct :ref:`gRPC <grpc>` endpoint (instead of JSON API)

> To get started quickly with the recommended application architecture clone the```

Reviewer: ```suggestion
To get started quickly with the recommended application architecture clone the
```

``create-daml-app`` application template:

.. code-block:: bash

  git clone https://github.com/digital-asset/create-daml-app

> ``create-daml-app`` is a small, but fully functional demo application implementing the recommended

Reviewer 2: Maybe worth pointing out that it’s not just a demo application but also a template for building your own apps.
EDIT: I see that you do this below but I would move it here to the beginning ,something like a “small but fully functional demo application intended to be used as a starting point for your own applications”

Reviewer: Some more options:
- is a fully functional demo application that provides the necessary skeleton for your own project.
- is a fully functional demo application that provides the necessary code base for your own project

Author: ok, I changed it in this spirit.

architecture, including

- DAML React libraries
- quick iteration against the :ref:`DAML Ledger Sandbox <sandbox-manual>`.

> - authentication

Reviewer: Maybe point to the "Authentication" section on this page? I think it would work better than pointing directly to the authentication page itself

Author: not sure if I find a link just pointing to a paragraph within the same page to useful. I find it almost more annoying to read because I have to check where the link is going.

Frontend
~~~~~~~~

We recommended building your frontend with the `React <https://reactjs.org>`_ framework. We provide
several support libraries to build your React frontend for a DAML application.

+--------------------------------------------------------------+--------------------------------------------------------------------------+
| Name                                                         | Summary                                                                  |
+==============================================================+==========================================================================+
| `@daml/react <https://www.npmjs.com/package/@daml/react>`_   | React hooks to query/create/exercise DAML contracts                      |
+--------------------------------------------------------------+--------------------------------------------------------------------------+
| `@daml/ledger <https://www.npmjs.com/package/@daml/ledger>`_ | DAML ledger object to connect and directly submit commands to the ledger |
+--------------------------------------------------------------+--------------------------------------------------------------------------+

> | `@daml/types  <https://www.npmjs.com/package/@daml/types>`_  | DAML data types                                                          |

Reviewer 2: Maybe “Typescript equivalents for DAML data types” or something like that would be clearer.

Reviewer: Another option "DAML data types represented in TypeScript"

Reviewer: Also update the description at npmjs.com to something like "Library for generating DAML data types in TypeScript. If you are writing a full-stack DAML application you will need it in order to be able to communicate between your React components and DAML code"

Would be good to explain there why this is needed

Author: @daml/types is not a library to generate data types in typescript, it's equivalents of core DAML types in typescript. Maybe it shouldn't be listed here at all because it's more of an implementation detail of the other two libraries. It would rather belong to a page describing our javascript bindings for people that want to use a different javascript frontend framework than react.

Reviewer: If it's a detail and users don't used it then you're absolutely right: they don't need to know about it 😄As we mention it later

"If you choose a different Javascript based frontend framework, the packages ``@daml/ledger``,
``@daml/types`` and the generated ``@daml2ts`` package provide you with the necessary interface code to connect and issue commands against your ledger."

a short clarification at npmjs.com would be good enough. Or we just don't mention it in the referenced sentence 😄

Author: I removed it and will reintroduce it in a separate javascript bindings page.

+--------------------------------------------------------------+--------------------------------------------------------------------------+

> You can install any of these libraries by running ``yarn add @daml/react`` in the ``ui`` directory

Reviewer: I would need to run a separate command for each library, right? So yarn add @daml/react, yarn add @daml/ledger and yarn add @daml/types. If that's the case then maybe write something like

"You can install these libraries by running the following command in the ``ui``directory of your project ``yarn add @``and the name of the library (e.g., ``yarn add @daml/react`` for installing the  ``@daml/react``library)"

Author: changed.

of your project. Please explore the ``create-daml-app`` example project to see the usage of these
libraries.

> The interface between the frontend and the backend of your application is defined by your DAML code.

Reviewer: This makes it sound like DAML is some sort of an interface or middleware language.  Why not move it to the backend side in the docs and explain everything there? If you run the sandbox it runs with your dar file. IMHO makes it easier to understand.

Author: Hmm, I think it kind of is an interface language. It tells you how data is represented and how it can be changed and by whom. I don't want to move this to the backend, because otherwise I'd have to talk about the frontend in the backend section.

Reviewer: @bame-da @hurryabit Can we have a consensus here on how we talk about DAML with respect to frontend and backend? At the moment in the recommended architecture it's represented as an interface language. If that's the way to go, fine. I just think it's underselling what it is and should reside on the backend side.

Author: I have changed the diagram and this sentence so that hopefully everybody is happy.

Author: added final changes by @nemanja-da.

To make your life easy when interacting with the ledger, the DAML assistant can generate
corresponding typescript data definitions for the data types declared in DAML.

.. code-block:: bash

  daml codegen ts .daml/dist/<your-project-name.dar> -o daml-ts

This will create a typescript project under ``daml-ts`` that you can install by changing directory
to ``daml-ts`` and running ``yarn install`` and then in the root of your project ``yarn workspaces
run build``. To start the DAML sandbox, run the ``daml-start.sh`` script at the root of the project

> directory of ``create-daml-app``.

Reviewer: ```suggestion
The codegen will generate a typescript project in the ``daml-ts``folder that needs to be connected with the rest of the architecture. To do so navigate to ``daml-ts`` and run ``yarn install`` and then ``yarn workspaces run build``.  We are currently working on making this step easier so that you don't have to do it manually.

Next, to start the DAML sandbox just run the ``daml-start.sh`` script in the ``create-daml-app`` folder.
```

Would try to explain why all the yarn commands for daml-ts project are necessary.

Author: added this suggestion with slight modifications.

If you choose a different Javascript based frontend framework, the packages ``@daml/ledger``,
``@daml/types`` and the generated ``@daml2ts`` package provide you with the necessary interface code
to connect and issue commands against your ledger.

Backend
~~~~~~~

The backend for your application can be any DAML ledger implementation. For applications where

> performance is not of outmost importance, we recommend using the :ref:`DAML JSON API <json-api>` as

Reviewer 2: I would leave out the statement about performance here. I think we don’t really have the numbers to back up that the JSON API gives you a significant performance hit and it can even be an improvement due to caching and the fact that it handles queries smarter than you might handle them yourself.

an interface. It is served by the HTTP JSON API server connected to the ledger API server. It
provides simple HTTP endpoints to interact with the ledger via GET/POST requests. If performance
does become a concern for you application, you can drop the HTTP JSON API server and use the
:ref:`gRPC API <grpc>` directly.

> When you use the ``create-daml-app`` template application you can start a local sandbox together

Reviewer 2: ```suggestion
When you use the ``create-daml-app`` template application, you can start a local sandbox together
```

with a JSON API server by running

.. code-block:: bash

> 

Reviewer 2: I think the docs on event-driven applications and state-driven applications are still relevant and useful. They are mostly independent of your client stack (unless your client stack forces you into one). E.g., I’ve had discussions around whether triggers should be state or event-based fairly frequently. Not quite sure where to integrate it but it would be a shame to drop it completely.

Author: I'm still quite undecided about the overall structure of the text, I think this part may survive, but maybe closer somewhere to the triggers?

Reviewer 2: The trigger API pretty much forces you to write state-based automation. It probably makes more sense to have something a bit more generic that talks about ledger clients in general.

Reviewer: My 2 cents are that this page should be just about the recommended architecture, straightforward and simple. Stat vs. event driven application discussion should be placed somewhere else (with concrete examples, current text in the docs reads like theory more).

Author: I also think this part needs to go somewhere else. It's way too scary for someone who just gets started.

Reviewer 2: I fully agree! I just would like for it to not be completely lost.

Author: I'll move it to a .back file for now and we figure out what to do with it later. Ok?

Reviewer 2: :+1:

example https://projectdabl.com offer an integrated authentication service, but you can also use an
external service provider for authentication like https://auth0.com. The DAML react libraries
support interfacing with an authenticated DAML ledger. Simply initialize your ``DamlLedger`` object
with the token obtained by an authentication service. How authentication works and the form of the
required tokens is described in the :ref:`Authentication <authentication>` section.

Developer workflow
~~~~~~~~~~~~~~~~~~

The DAML SDK enables a local development environment with fast iteration cycles. If you run
``daml-reload-on-change.sh`` of the ``create-daml-app``, a local DAML sandbox ledger is started that
is updated with your most recent DAML code on any change. Next, you can start your frontend in
development mode by changing to your ``ui`` directory and run ``yarn start``. This will reload your
frontend whenever you make changes to it. You can add unit tests for your DAML models by writing
:ref:`DAML scenarios <testing-using-scenarios>`. These will also be reevaluated on change.  A
typical DAML developer workflow is to

  #. Make a small change to your DAML data model

>   #. Optionally test your DAML code and with scenarios (add reference to scenario testing in the docs)

Reviewer: ```suggestion
  #. Optionally test your DAML code with scenarios 
```
Can you add a reference for scenario testing pointing to the appropriate docs page? That should be the very last thing 😄

Author: yes, that's a good idea.
